File: blocksandstuff-blocks/src/main/java/org/everbuild/blocksandstuff/blocks/placement/RotatedPillarPlacementRule.java
package org.everbuild.blocksandstuff.blocks.placement;

import net.minestom.server.instance.block.Block;
import net.minestom.server.instance.block.rule.BlockPlacementRule;
import org.everbuild.blocksandstuff.common.utils.DirectionUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

public class RotatedPillarPlacementRule extends BlockPlacementRule {
    public RotatedPillarPlacementRule(@NotNull Block block) {
        super(block);
    }

    @Override
    public @Nullable Block blockPlace(@NotNull PlacementState placementState) {
        if (placementState.playerPosition() == null) return placementState.block();
        return placementState.block()
                .withProperty(
                        "axis",
                        DirectionUtils.getAxis(
                                Objects.requireNonNull(placementState.blockFace())
                                        .toDirection()
                        )
                );
    }

}

File: blocksandstuff-blocks/src/main/kotlin/org/everbuild/blocksandstuff/blocks/placement/RotatedPillarPlacementRule.kt
package org.everbuild.blocksandstuff.blocks.placement

import net.minestom.server.instance.block.Block
import net.minestom.server.instance.block.rule.BlockPlacementRule
import org.everbuild.blocksandstuff.common.utils.axis

class RotatedPillarPlacementRule(block: Block) : BlockPlacementRule(block) {
    override fun blockPlace(placementState: PlacementState): Block {
        if (placementState.playerPosition() == null) return placementState.block()
        return placementState.block()
            .withProperty("axis", placementState.blockFace()!!.toDirection().axis)
    }
}
